ogre2/src/Ogre2MeshFactory.cc

```diff
@@ -88,6 +89,17 @@ void Ogre2MeshFactory::Clear()
   this->ogreMeshes.clear();
 }
 
+void Ogre2MeshFactory::ClearMaterialsCache()
```

Comment:
I verified that this fixes the memory leak. A different issue is created after removing these materials though. Here are the steps to reproduce:
1. Insert a model like this one (I just dragged and dropped) https://app.ignitionrobotics.org/GoogleResearch/fuel/models/Weisshai_Great_White_Shark
    * it needs to be a model that references a texture in the dae or obj file but does not have a <material> sdf element
 1. Insert the same model again -> model appear grey 

After debugging a little, I remembered that these materials were created as "templates". Later on in this line https://github.com/ignitionrobotics/ign-rendering/blob/6fd0c03ce32568f8bc91c1041fc169cc7622248f/ogre2/src/Ogre2MeshFactory.cc#L601 a "new instance" is created and assigned to the submesh. This is done by looking for the name of the template material stored in the submesh, getting the template material from the scene, then assigning a copy to the submesh. Now that this template material is removed, subsequent meshes requiring the same material will not be able to find it. 

One possible solution is to remove the code for creating these template materials, i.e. the code here: https://github.com/ignitionrobotics/ign-rendering/blob/6fd0c03ce32568f8bc91c1041fc169cc7622248f/ogre2/src/Ogre2MeshFactory.cc#L455-L471 

we just then need to always create a new material and assign it to the submesh, i.e. replace the logic here: https://github.com/ignitionrobotics/ign-rendering/blob/6fd0c03ce32568f8bc91c1041fc169cc7622248f/ogre2/src/Ogre2MeshFactory.cc#L590-L602


I ran out of time today and haven't found a way to do his cleanly yet..

Reply:
@iche033, I don't fully understand why this is happening, I would expect this three things

 - when we load a new mesh then the "template" material is loaded and removed.
 - When we remove the mesh we remove the material created fom the template
 - At this point we should not have any material associated with the mesh in memory.
 - Finally we load again the same mesh and I would expect to reload everything again (template and material created from the template).

am I missing something ?

Reply:
> Finally we load again the same mesh and I would expect to reload everything again (template and material created from the template).

we have this check here to skip loading the mesh if it's already loaded in ogre: https://github.com/ignitionrobotics/ign-rendering/blob/6fd0c03ce32568f8bc91c1041fc169cc7622248f/ogre2/src/Ogre2MeshFactory.cc#L167

That was mainly done for performance reason, e.g. when launching a world with hundreds of the same model, esp if they have large complex meshes.